File: software/examples/rust/example_interrupt.rs
use std::{io, error::Error};
use std::thread;
use tinkerforge::{ip_connection::IpConnection, 
                  io4_bricklet::*};


const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your IO-4 Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let io = Io4Bricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
    // Don't use device before ipcon is connected.

     let interrupt_receiver = io.get_interrupt_callback_receiver();

        // Spawn thread to handle received callback messages. 
        // This thread ends when the `io` object
        // is dropped, so there is no need for manual cleanup.
        thread::spawn(move || {
            for interrupt in interrupt_receiver {           
                		println!("Interrupt Mask: {:b}", interrupt.interrupt_mask);
		println!("Value Mask: {:b}", interrupt.value_mask);
		println!();
            }
        });

		// Enable interrupt on pin 0
		io.set_interrupt(1 << 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_interrupt.go
package main

import (
	"fmt"
	"tinkerforge/io4_bricklet"
	"tinkerforge/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your IO-4 Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	io, _ := io4_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	io.RegisterInterruptCallback(func(interruptMask uint8, valueMask uint8) {
		fmt.Printf("Interrupt Mask: %b\n", interruptMask)
		fmt.Printf("Value Mask: %b\n", valueMask)
		fmt.Println()
	})

	// Enable interrupt on pin 0
	io.SetInterrupt(1 << 0)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
